pkg/rc/replication_controller.go

```diff
@@ -0,0 +1,258 @@
+	// The goroutine continues running until a value is received on the quit channel.
+	// The error channel will be closed in response.
+	WatchDesires(quit <-chan struct{}) <-chan error
+}
+
+// I just copied this from kp.Store because it was a huge interface and I really only care about these methods!
+type kpStore interface {
+	SetPod(key string, manifest pods.Manifest) (time.Duration, error)
+	DeletePod(key string) (time.Duration, error)
+}
+
+type replicationController struct {
+	rcstore.ReplicationControllerData
+
+	logger logging.Logger
+
+	kpStore       kpStore
+	scheduler     Scheduler
+	podApplicator labels.Applicator
+}
+
+// replicationController implements the ReplicationController interface.
+var _ ReplicationController = &replicationController{}
+
+func New(
+	rcData rcstore.ReplicationControllerData,
+	kpStore kpStore,
+	scheduler Scheduler,
+	podApplicator labels.Applicator,
+) *replicationController {
+	return &replicationController{
+		ReplicationControllerData: rcData,
+
+		logger:        logging.DefaultLogger,
+		kpStore:       kpStore,
+		scheduler:     scheduler,
+		podApplicator: podApplicator,
+	}
+}
+
+func (rc *replicationController) WatchDesires(quit <-chan struct{}) <-chan error {
+	desiresChanged, errInChannel := rc.ReplicationControllerData.Watch(quit)
+
+	errOutChannel := make(chan error)
+	channelsClosed := make(chan struct{})
+
+	// When seeing any changes, try to meet then.
+	// If meeting produces any error, send it on the output error channel.
+	go func() {
+		for _ = range desiresChanged {
+			err := rc.meet()
+			if err != nil {
+				rc.logger.NoFields().Errorf("Could not meet: %s", err)
+				errOutChannel <- err
+			}
+		}
+		channelsClosed <- struct{}{}
+	}()
+
+	// When seeing any errors, forward them to the output error channel.
+	// (Notice that two goroutines may be writing to the output error channel)
+	go func() {
+		for err := range errInChannel {
+			errOutChannel <- err
+		}
+		channelsClosed <- struct{}{}
+	}()
+
+	// If both channels I'm watching are closed,
+	// which would happen when my caller requests a quit
+	// (the quit gets forwarded to the rc watcher)
+	// close my output channel.
+	go func() {
+		<-channelsClosed
+		<-channelsClosed
+		close(channelsClosed)
+		close(errOutChannel)
+	}()
+
+	return errOutChannel
+}
+
+func (rc *replicationController) meet() error {
+	rc.logger.NoFields().Infof("Meeting with desired replicas %d, disabled %v", rc.DesiredReplicas(), rc.Disabled())
+
+	// If we're disabled, we do nothing, nor is it an error
+	// (it's a normal possibility to be disabled)
+	if rc.Disabled() {
+		return nil
+	}
+
+	eligible, err := rc.eligibleNodes()
+	if err != nil {
+		return err
+	}
+	rc.logger.NoFields().Infof("Eligible to be on nodes %s", eligible)
+
+	current, err := rc.CurrentNodes()
+	if err != nil {
+		return err
+	}
+	rc.logger.NoFields().Infof("Currently on nodes %s", current)
+
+	if rc.DesiredReplicas() > len(current) {
+		possible := difference(eligible, current)
```

Comment:
I'd leave some kind of scheduler to do here. This code currently assumes that the label selector - existing nodes currently used by the pod is a sufficient decision making process for whether an app should run. Note that Docker/runc will be free of the constraint of not running more than once on a node. I'd like to acknowledge that in this code.